### PR TITLE
Fix warning "FutureWarning: ArrowStringArrayNumpySemantics._data is a deprecated" on Pandas conversion

### LINKED ISSUE #9182: FutureWarning: ArrowStringArrayNumpySemantics._data is a deprecated and will be removed in a future version, use ._pa_array instead
### What happens?

The line:
```python
self._con.register(name, df)
```
Is giving me the above warning:
```python
FutureWarning: ArrowStringArrayNumpySemantics._data is a deprecated and will be removed in a future version, use ._pa_array instead
```
```self._con``` is initialized in my __init__  method:

```python
        self._con = duckdb.connect(database=':memory:')
```

The snippet below shows how the dataframe (df) is constructed:

```python

        dfs = {
            'EUROPE': pd.DataFrame({'EUROPE': list(local.EU)}),
            'AMERICAS': pd.DataFrame({'AMERICAS': list(local.AMERICAS)}),
        }
        for name, df in dfs.items():
            self._con.register(name, df)
```

### To Reproduce

```python

        dfs = {
            'EUROPE': pd.DataFrame({'EUROPE': list(local.EU)}),
            'AMERICAS': pd.DataFrame({'AMERICAS': list(local.AMERICAS)}),
        }
        for name, df in dfs.items():
            self._con.register(name, df)
```

I show local.EU as an example of the list used to construct the dataframe:

```python
EU: final = {
    'BELGIUM',
    'BULGARIA',
    'DENMARK',
    'FINLAND',
    'FRANCE',
    'GERMANY',
    'GREECE',
    'ROMANIA',
    'SLOVAKIA',
 }
```

### OS:

win11, Pandas 2.1.1 and python 3.11.5

### DuckDB Version:

0.9

### DuckDB Client:

python

### Full Name:

Dr Abed Hammoud

### Affiliation:

Alant Health

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by pdet**
Fixes: #9182

### CODE REVIEW

**Comment by Tishj** [comment]
I have no clue of the implications of the change from `data` to `_pa_array`.

Seeing as this is Python this could be a world of difference, I can't find any docs for `_pa_array`

**Comment by pdet** [comment]
AFAIK, the `future.infer_string` will force pandas to use the `pyarrow_numpy` storage for strings, the code simply checks `_pa_array` is available (which indicates that it is a `pyarrow_numpy`) and uses that.

Indeed, I guess docs are a bit shady, you need to go to some discussion lists and so on to find these.

I just noticed I do not check the actual query result, I'll add that to the test as well.

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/tests/fast/pandas/test_raise_warning.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+from typing import final
+import pytest
+
+
+def test_9182():
+    pd.set_option('future.infer_string', True)
+
+    _con = duckdb.connect(database=':memory:')
+
+    EU: final = {
```
[the remark this answers is not in the document]
Comment (pdet) [reply tools/pythonpkg/tests/fast/pandas/test_raise_warning.py:12]: That's just copy-pasted from the issue,  i'm also fine on removing that.

**Comment by Tishj** [comment]
@pdet

Doesn't that mean that it's an arrow array and `_pa_array` will convert it into a numpy array?

File: tools/pythonpkg/tests/fast/pandas/test_raise_warning.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import duckdb
+import pandas as pd
```
Comment (Tishj) [inline tools/pythonpkg/tests/fast/pandas/test_raise_warning.py:2]: This needs `pd = pytest.importorskip("pandas")

Comment (pdet) [reply tools/pythonpkg/tests/fast/pandas/test_raise_warning.py:2]: Oh we removed pandas from the test dependencies? Was not aware of that, will change it to importorskip

**Comment by pdet** [comment]
> ```python
> import pandas as pd
> 
> pd.set_option('future.infer_string', True)
> 
> df = df = pd.DataFrame({'a': ['a', 'b', 'c']})
> df['a'].__class__
> df['a'].array
> ```
> 
> ```
> >>> df['a'].__class__
> <class 'pandas.core.series.Series'>
> 
> >>> df['a'].array
> <ArrowStringArrayNumpySemantics>
> ['a', 'b', 'c']
> Length: 3, dtype: string
> ```
> 
> ^ I don't even know what this is, it's not possible to call `help()` on it
> 
> It's apparently this:
> 
> ```
> >>> df['a'].array.__class__
> <class 'pandas.core.arrays.string_arrow.ArrowStringArrayNumpySemantics'>
> ```
> 
> This works:
> 
> ```python
> help(pd.core.arrays.string_arrow.ArrowStringArrayNumpySemantics)
> ```
> 
> ```python
> >>> df['a'].array._pa_array
> <pyarrow.lib.ChunkedArray object at 0x1622cd620>
> [
>   [
>     "a",
>     "b",
>     "c"
>   ]
> ]
> ```
> 
> So it's actually the internal pyarrow array 👀 I think that means we will implicitly create a numpy array out of it this way
> 
> This actually sounds like it would be a good opportunity to put into practice where I prepared for a while ago:
> 
> ```c++
> 	switch (backend) {
> 	case PandasColumnBackend::NUMPY: {
> 		NumpyScan::Scan(bind_data, count, offset, out);
> 		break;
> 	}
> 	default: {
> 		throw NotImplementedException("Type not implemented for PandasColumnBackend");
> 	}
> 	}
> ```
> 
> This should ideally be a `PandasColumnBackend::ARROW` (we don't have that yet)

Fair enough, I can have a go at that.

**Comment by Tishj** [comment]
> Fair enough, I can have a go at that.

Cool, if you run into too much trouble it's probably fine to leave it like this for now
But I think it's good to be aware that we're converting an arbitrarily sized pyarrow array into numpy here

**Comment by Tishj** [comment]
@pdet Also another step we can take first is to fall back to an arrow scan if all the arrays are PyArrowNumpy arrays, we also do that for Pyarrow backed pandas arrays

**Comment by pdet** [comment]
Closing this PR for now, since the scope is way bigger than I initially planned, will reopen when it is done.

**Pull request closed without merging**